include/Basic/SourceManager.hpp

```diff
@@ -0,0 +1,273 @@
+
+    /// The content cache of the file.
+    ContentCache const *_contentCache;
+
+public:
+    FileInfo()
+        : _importLoc(0)
+        , _contentCache(nullptr) {};
+
+    /// Create a FileInfo object with the given SourceLocation and ContentCache.
+    static FileInfo
+    get(SourceLocation includeLoc, ContentCache &contentCache,
+        llvm::StringRef fileName)
+    {
+        FileInfo info;
+
+        info._importLoc = includeLoc;
+        info._contentCache = &contentCache;
+        contentCache.setFileName(fileName);
+        return info;
+    }
+
+    SourceLocation getImportLoc() const
+    {
+        return _importLoc;
+    }
+
+    ContentCache const &getContentCache() const
+    {
+        return *_contentCache;
+    }
+
+    llvm::StringRef getFileName() const
+    {
+        return _contentCache->getFileName();
+    }
+};
+
+///
+/// @class SourceLocEntry
+/// @brief Represents a specific location in the source code.
+///
+/// A SourceLocEntry is an object that represents a specific location
+/// in the source code. It is used to refer to a specific character in a
+/// specific file. The SourceManager class is responsible for creating and
+/// interpreting SourceLocEntry objects.
+///
+/// A SourceLocEntry is basicly an offset into the complete source code. The
+/// SourceManager knowings to which file this offset belongs can interpret it
+/// and provide useful informations from it.
+///
+class SourceLocEntry {
+
+private:
+    /// The offset of the source location.
+    uint32_t _offset;
+
+    /// The file information for the source location.
+    FileInfo _file;
+
+public:
+    SourceLocEntry(uint32_t offset, FileInfo const &file)
+        : _offset(offset)
+        , _file(file)
+    {
+    }
+
+    /// Returnn a const reference to the FileInfo object.
+    FileInfo const &getFileInfo() const
+    {
+        return const_cast<SourceLocEntry *>(this)->getFile();
+    }
+
+    FileInfo &getFile()
+    {
+        return _file;
+    }
+
+    /// Return the offset of the source location.
+    uint32_t getOffset() const
+    {
+        return _offset;
+    }
+};
+
+///
+/// @class SourceManager
+/// @brief Manages the source code of a glu program.
+///
+/// The SourceManager class is responsible for managing the source code of a glu
+/// program. For now, it is responsible for loading files, caching their
+/// content, and for providing information about the source code. At term, the
+/// FileManager will be responsible of the file system and the SourceManager
+/// will be responsible for the source code.
+///
+/// The SourceManager knows how to interpret a SourceLocation object and provide
+/// information about the source code at that location in an efficient way.
+///
+class SourceManager {
+
+private:
+    /// Every source location of the source code.
+    llvm::SmallVector<SourceLocEntry, 0> _sourceLocs;
+    /// The content caches of the loaded files.
+    std::vector<ContentCache *> _contentCaches;
```

Comment:
why
they're already in the above vector, why do you need a second vector?
+ it's using std vector